lib/pure/random.nim

```diff
@@ -617,15 +622,60 @@ proc shuffle*[T](x: var openArray[T]) =
 
   shuffle(state, x)
 
-when not defined(nimscript) and not defined(standalone):
-  import times
+when not defined(standalone):
+  when defined(nimscript):
+    import std/hashes
+
+    var baseState = block:
+      var ret = Rand(
+        a0: CompileTime.hash.Ui,
+        a1: CompileDate.hash.Ui)
+      if not ret.isValid:
+        ret = DefaultRandSeed
+      ret
+  elif defined(js):
+    import std/times
+  else:
+    import std/[hashes, os, sysrand, monotimes]
+
+    var baseState: Rand = block:
+      var
+        ret: Rand
+        urand: array[sizeof(Rand), byte]
+
+      for i in 0 .. 7:
+        if sysrand.urandom(urand):
+          copyMem(ret.addr, urand[0].addr, sizeof(Rand))
+          if ret.isValid:
+            break
+
+      if not ret.isValid:
+        # When 2 processes executed at same time on different machines,
+        # `ret` can still have same value.
+        let
+          pid = getCurrentProcessId()
+          t = getMonoTime().ticks
+        ret.a0 = pid.hash().Ui
+        ret.a1 = t.hash().Ui
+        if not ret.isValid:
+          ret.a0 = pid.Ui
+          ret.a1 = t.Ui
```

Comment:
When would the data supplied by `sysrand.urandom` not correctly initialize the global random state?

Reply:
xoroshiro128+ doesn't work when all bits of state are zero.
If `sysrand.urandom` is not guaranteed to never return zeros when sizeof(Rand) bytes are taken from it, we need to check returned value.
Maybe there is a bug in OS and `sysrand.urandom` returns zero.
Even if `sysrand.urandom` is working correctly, you might get a zero unfortunately (probability of 128bits random number returns 0 is 1/2^128 ).

According to the comment of `sysrand.urandom` proc,
> If the call succeeds, returns `true`.

In other words, it can fail. So we need alternative ways to get random number.
Even if probability of `sysrand.urandom` returns 128bit zeros is tiny, it is easy to take care of such case if we already have alternative ways to get random number.

Reply:
What about `AND`ing the state with `1` after seeding via urandom? That would ensure that the state is never zero.
Alternately, an explicit check and retry would work.

Reply:
`isValid(r: Rand): bool` func checks whether state is zero or not.
```nim
func isValid(r: Rand): bool {.inline.} =
  not (r.a0 == 0 and r.a1 == 0)
```
This PR retrys `urandom` if `isValid` was false.